zerver/lib/bugdown/help_settings_links.py

```diff
@@ -0,0 +1,86 @@
+import re
+import os
+import markdown
+import six
+from typing import Any, Dict, Iterable, List, MutableSequence, Optional, Tuple, Union, Text
+from typing.re import Match
+from xml.etree.cElementTree import Element
+from markdown.preprocessors import Preprocessor
+
+REGEXP = re.compile(r'\{setting\((?P<setting_name>.*?)\)\}')
+
+link_mapping = {
+    # a mapping from the setting name as it appears in the user interface to
+    # the name of its setting type and a relative link that can be used to get
+    # to that setting
+    'Your account': ['Settings', '/#settings/your-account'],
+    'Custom emoji': ['Manage organization', '/#organization/emoji-settings'],
+}
+
+settings_markdown = """
+1. From your desktop, click on the **gear**
+   (<i class="icon-vector-cog"></i>) in the upper right corner.
+
+1. Select **%s**.
+
+1. On the left, click %s.
+"""
+
+
+class SettingHelpExtension(markdown.Extension):
+
+    def __init__(self, *args, relative_links=False, **kwargs):
+        self.relative_links = relative_links
+
+    def extendMarkdown(self, md: markdown.Markdown, md_globals: Dict[str, Any]) -> None:
+        """ Add SettingHelpExtension to the Markdown instance. """
+        md.registerExtension(self)
+        print(md.preprocessors.__dict__)
+        md.preprocessors.add('setting', Setting(relative_links=self.relative_links),
+                             '_begin')
+
+
+class Setting(Preprocessor):
+    def __init__(self, *args, relative_links: bool, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.relative_links = relative_links
+
+    def run(self, lines: List[str]) -> List[str]:
+        done = False
+        while not done:
+            for line in lines:
+                loc = lines.index(line)
+                match = REGEXP.search(line)
+
+                if match:
+                    text = [self.handleMatch(match)]
+                    # The line that contains the directive to include the macro
+                    # may be preceded or followed by text or tags, in that case
+                    # we need to make sure that any preceding or following text
+                    # stays the same.
+                    line_split = REGEXP.split(line, maxsplit=0)
+                    preceding = line_split[0]
+                    following = line_split[-1]
+                    text = [preceding] + text + [following]
+                    lines = lines[:loc] + text + lines[loc+1:]
+                    break
+            else:
+                done = True
+        return lines
+
+    def handleMatch(self, match: Match[Text]) -> Optional[Element]:
+        setting_name = match.group('setting_name')
+        setting_type_name = link_mapping[setting_name][0]
+        setting_link = link_mapping[setting_name][1]
+        if self.relative_links:
+            setting_reference = "[%s](%s)" % (setting_name, setting_link)
+        else:
+            setting_reference = "**%s**" % (setting_name,)
```

Comment:
This leaves in the step-by-step instructions on how to get to the setting, even if we also include the link. See screenshot.
![screenshot from 2018-04-18 23-16-00](https://user-images.githubusercontent.com/549661/38969486-902497b4-435e-11e8-9490-c527204fc84d.png)

@rishig I like leaving in the detailed instructions just because it is more verbose, and it allows users to see where we keep the setting so that if they want to find it for future reference, they can. (rather than depending on the link to find the setting whenever they need it). However, I could see that more concise documentation might be better. What do you think?

Reply:
Also, also, I don't know how to compile these without having 'html_settings_links'=True, so I'm just relying on the code looking right to verify it's working. That strategy always makes me nervous.

Reply:
@rishig These comments still apply.

Reply:
oops, thought I had replied! 
* Leaving the detailed instructions sounds good; easy to change later if we change our minds
* No link for "Manage organization"
* Not sure about the compiling question.

Reply:
For without `html_settings_links=True`, I think what we'll want is a unit test.  It's a bit hard to explain how to do that, so I'm tempted to just show by example (aka write the test when I merge this).

But for you to manually verify, you can always just set the variable to false and reload the page.